src/main/java/edu/byu/cs/autograder/git/specs.md

```diff
@@ -0,0 +1,37 @@
+
+## Recommendation
+Based on all the information above, I propose the following:
+1. We count only commits towards the commit limit if they were authored after **the most recent timestamp of the commit of any commit that resulted in an accepted submission from any phase**.
+2. We throw an error and refuse to grade if we detect any commits that are authored before their parent commit (indicates intentionally messing with the local clock, or cherry-picking commits).
+3. We withhold the score as “not qualified” if any of the commits are authored in the future of the server timestamp (indicates messing with the local clock, or it could be a simple mistake).
+4. We explicitly preserve a student’s commit verification result for any subsequent submissions on the same phase. I.E. If they pass on their first valid submission, then we’ll continue passing them even if they are resubmitting to fix quality errors etc… And, if they fail on their first submissions, they’ll be able to continue resubmitting, but none of those grades will be submitted to Canvas even if they are resolving quality errors.
+
+### Advantages
+1. A student is able to submit multiple times on the autograder with code that fails, and are only held accountable for meeting commit requirements on their first successful submission.
+2. A student cannot cheese the system by quickly adding nearly-empty commits after a successful run in order to resubmit without a penalty
+3. This definition handles the cases where students don’t submit to the autograder in between phase submissions.
+4. It is not vulnerable to simply rebasing commits in order to count new commits.
+5. It also protects against simply changing the local clock to count commits as older.
+6. Additionally, a student who has their clock set in the future will still be held accountable for authoring new commits.
+7. Students cannot resubmit their code to the autograder with an earlier passing commit in order to count more commits as credit for the current phase. If we only evaluated the most recent passing submission, then a student could push their maximum time back by resubmitting with an earlier commit that still passes. Even if the commit lowers results in a lower score (from quality checks etc…), this wouldn't hurt them since we don’t send lower scores to Canvas.
+
+### Shortcomings
+1. It does not protect against a student modifying **all the commit timestamps** of their code block by a fixed amount, but if a student is familiar enough with git to do such an advanced operation, they probably deserve to keep their points gained.
+2. Additionally, any student who was not intentionally breaking the system but accidentally has their clock set forward for some reason (I’ve done that before) would be prevented from having their submission graded.
+3. Students who cherry-pick commits out of chronological order will be prevented from grading.
+4. This still assumes that students submit the phases in order (probably safe).
```

Comment:
I agree this is probably safe. I believe that we've discussed enforcing this formally, but have hesitated doing so because students have not been informed of this policy. Again, I think this will require clear and deliberate communicate with the students as an accompanying change should be made to make this be formally enforced

Reply:
The students will either figure it out quickly, or they'll talk to the TA's and not lose any points in the end. I think this is a reasonable assumption to make.